mqtt/mqttd/src/broker/mod.rs

```diff
@@ -5,45 +5,107 @@ mod snapshot;
-};
+use futures_util::future::{select, select_all, Either};
+use tracing::{error, info};
+
+use mqtt_broker::{FilePersistor, Message, Persist, SystemEvent, VersionedFileFormat};
+
+use crate::broker::snapshot::start_snapshotter;
 
 pub async fn run<P>(config_path: Option<P>) -> Result<()>
 where
     P: AsRef<Path>,
 {
-    let config = bootstrap::config(config_path).context(LoadConfigurationError)?;
+    let settings = bootstrap::config(config_path).context(LoadConfigurationError)?;
+    let listener_settings = settings.listener().clone();
 
     info!("loading state...");
-    let persistence_config = config.broker().persistence();
+    let persistence_config = settings.broker().persistence();
     let state_dir = persistence_config.file_path();
 
     fs::create_dir_all(state_dir.clone())?;
     let mut persistor = FilePersistor::new(state_dir, VersionedFileFormat::default());
     let state = persistor.load().await?;
     info!("state loaded.");
 
-    let broker = bootstrap::broker(config.broker(), state).await?;
+    let broker = bootstrap::broker(settings.broker(), state).await?;
+    let mut broker_handle = broker.handle();
 
-    info!("starting snapshotter...");
     let snapshot_interval = persistence_config.time_interval();
     let (mut snapshotter_shutdown_handle, snapshotter_join_handle) =
         start_snapshotter(broker.handle(), persistor, snapshot_interval).await;
 
-    let shutdown = shutdown::shutdown();
-    pin_mut!(shutdown);
-
-    info!("starting server...");
-    let state = bootstrap::start_server(config, broker, shutdown).await?;
+    let shutdown_signal = shutdown::shutdown();
+
+    // start broker
+    let server_join_handle =
+        tokio::spawn(bootstrap::start_server(settings, broker, shutdown_signal));
+
+    // start sidecars
+    // TODO REVIEW: it is fine to blow up here
+    //              a failure here would indicate either bridge or command handler init failed
+    //              broker doesn't open ports until these are both initialized
+    let state;
+    if let Some((sidecars_shutdown, sidecar_join_handles)) =
+        bootstrap::start_sidecars(broker_handle.clone(), listener_settings).await?
+    {
+        // combine future for all sidecars
+        // wait on future for sidecars or broker
+        // if one of them exits then shut the other down
+        // TODO REVIEW: we are only blowing up if either:
+        //              1 - we can't stop the server and can't get the needed state
+        //              2 - we can't signal server or sidecars to shutdown
+        let sidecars_fut = select_all(sidecar_join_handles);
+        state = match select(server_join_handle, sidecars_fut).await {
+            // server finished first
+            Either::Left((server_output, sidecar_join_handles)) => {
+                // extract state from finished server
+                let state = server_output??;
+
+                // shutdown sidecars
+                sidecars_shutdown.shutdown().await?;
```

Comment:
Just as an idea: will it improve the code if we group together code, that calls shutdown on each sidecar and waits until it finishes? I mean `fn start_sidecars` returns a struct that holds all joins and a shoutdown handles for each sidecar, and exposes `fn shutdown` method that calls shutdown for each sidecar and also joins all of them.

Reply:
I made a sidecar manager struct to do this